ev3dev2/motor.py

```diff
@@ -1921,6 +1954,96 @@ def on(self, left_speed, right_speed):
         self.left_motor.run_forever()
         self.right_motor.run_forever()
 
+    def follow_line(self,
+            kp, ki, kd,
+            speed, keep_following_function,
+            target_light_intensity=None,
+            follow_left_edge=True,
+            WHITE=60,
+            OFF_LINE_COUNT_MAX=20,
+            SLEEP_TIME=0.01
+        ):
+        """
+        PID line follower
+
+        ``kp``, ``ki``, and ``kd`` are the PID constants.
+
+        ``speed`` is the desired speed of the midpoint of the robot
+
+        ``keep_following_function`` is called to determine if we should keep
+            following the line or stop
+
+        ``target_light_intensity`` is the reflected light intensity when the color sensor
+            is on the edge of the line.  If this is None we assume that the color sensor
+            is on the edge of the line and will take a reading to set this variable.
+
+        ``follow_left_edge`` determines if we follow the left or right edge of the line
+
+        ``WHITE`` is the reflected_light_intensity that is used to determine if we have
+            lost the line
+
+        ``OFF_LINE_COUNT`` is how many consecutive times through the loop the
+            reflected_light_intensity must be greater than ``WHITE`` before we
+            declare the line lost and raise an exception
+
+        ``SLEEP_TIME`` is how many seconds we sleep on each pass through
+            the loop.  This is to give the robot a chance to react
+            to the new motor settings. This should be something small such
+            as 0.01 (10ms).
+        """
+        assert self.cs, "ColorSensor must be defined"
+
+        if target_light_intensity is None:
+            target_light_intensity = self.cs.reflected_light_intensity
+
+        integral = 0.0
+        last_error = 0.0
+        derivative = 0.0
+        off_line_count = 0
+        speed_native_units = speed.to_native_units(self.left_motor)
+        MAX_SPEED = SpeedNativeUnits(self.max_speed)
+
+        while keep_following_function(self):
+            reflected_light_intensity = self.cs.reflected_light_intensity
+            error = target_light_intensity - reflected_light_intensity
+            integral = integral + error
```

Comment:
Does the implementation you were referencing not have an accumulator cap? I'm a bit surprised but I guess if it works it's good. Probably if it accumulates too much it's lost the line anyway so it doesn't much matter.

Reply:
It did not, I mainly followed this video as a guide
https://www.youtube.com/watch?v=AMBWV_HGYj4&t=1127s